File: src/main/ed/js/Prototype.java
/**
*    Copyright (C) 2008 10gen Inc.
*
*    This program is free software: you can redistribute it and/or  modify
*    it under the terms of the GNU Affero General Public License, version 3,
*    as published by the Free Software Foundation.
*
*    This program is distributed in the hope that it will be useful,
*    but WITHOUT ANY WARRANTY; without even the implied warranty of
*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*    GNU Affero General Public License for more details.
*
*    You should have received a copy of the GNU Affero General Public License
*    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package ed.js;
import ed.log.*;
import ed.js.*;
import ed.js.engine.*;
import ed.js.func.*;
import ed.util.*;

/**  Methods borrowed from Prototype
 * @anonymous name : {Class.create} desc : {Used to create an psuedo-class in JavaScript, optionally setting the class's parent, prototype, constructor, and adding methods.}  param : {type : (function) name : (parent), desc : (parent class)}, param : { type : (Object), name : (methods), desc : (a collection of method name/body pairs to add to the class)} return : {type : (Class) desc : (the new class)}
 * @expose
 */
public class Prototype {

    /** Object.extend: copy properties from src to dest.
     *  @returns dest
     */
    public static class Object_extend extends JSFunctionCalls2 {
        public Object call( Scope s , Object dest , Object src , Object [] extra ){

            if ( dest == null )
                throw new RuntimeException( "Object.extends dest is null" );

            if ( src == null )
                throw new RuntimeException( "Object.extends src is null" );

            if ( ! ( dest instanceof JSObject ) )
                throw new RuntimeException( "Object.extends destination is not an object.  is a : " + dest.getClass() );

            if ( ! ( src instanceof JSObject ) )
                throw new RuntimeException( "Object.extends src is not an object.  is a : " + src.getClass() );

            JSObject jdest = (JSObject)dest;
            JSObject jsrc = (JSObject)src;
            for(String key : jsrc.keySet()){
                jdest.set(key, jsrc.get(key));
            }

            return dest;
        }
    }

    /** Object.keys: fetch all the property names from an object
     *  @returns a JavaScript Array, each element containing a key from the
     *  object
     */
    public static class Object_keys extends JSFunctionCalls1 {
        public Object call( Scope s , Object o , Object [] extra ){
            JSObject jo = (JSObject)o;
            JSArray jar = new JSArray();
            for(String key : jo.keySet()){
                jar.add(new JSString(key));
            }

            return jar;
        }
    }


    /** Object.values: fetch all the property values from an object
     *  @returns a JavaScript Array of the properties of the object
     */
    public static class Object_values extends JSFunctionCalls1 {
        public Object call( Scope s , Object o , Object [] extra ){
            JSObject jo = (JSObject)o;
            JSArray jar = new JSArray();
            for(String key : jo.keySet()){
                jar.add(jo.get(key));
            }

            return jar;
        }
    }

    /** $A: convert an object into an array
     *  prototype uses this all over, but I don't think we'll need it
    public static JSFunction dollarA = new JSFunctionCalls1(){
            public Object call( Scope s , Object iterable , Object [] extra ){
                if(iterable == null)
                    return new JSArray();
                JSObject jiterable = (JSObject)iterable;
                if(jiterable.get("toArray") != null){
                    JSObject jtoarray = (JSObject)jiterable.get("toArray");
                    if(jtoarray instanceof JSFunction){
                        JSFunction jfunc = (JSFunction)jtoarray;
                        return jfunc.call( s );
                    }
                }
                Object length = jiterable.get("length");
                int len = 0;
                if(length != null){
                    if(! (length instanceof Number) )
                        length = StringParseUtil.parseInt(length.toString(), 0);
                    len = ((Number)length).intValue();
                }

                JSArray ary = new JSArray(len);
                while (len-- > 0) ary.setInt(len, jiterable.get(len));
                return ary;
            }
        };
    */

    /** Function.bind: given a function and some arguments,
     *  return a new function that provides those arguments.
     *  i.e. g = f.bind(o, 1, 2);
     *  g(3) is now the same as f.call(o, 1, 2, 3).
     */
    public static class Function_bind extends JSFunctionCalls1 {

        public Object call( final Scope s , Object o , Object [] arguments ){

            final JSFunction f = (JSFunction)s.getThis();
            final Object t = o;
            final Object [] farguments = arguments;

            return new JSFunctionCalls0() {
                public Object call( Scope s , Object [] args ){
                    int totlen = 0;
                    if(farguments != null) totlen += farguments.length;
                    if(args != null) totlen += args.length;
                    Object [] total = new Object [totlen];

                    int j = 0; // where we are in total
                    if(farguments != null){
                        for(int i = 0; i < farguments.length; i++, j++){
                            total[j] = farguments[i];
                        }
                    }

                    if(args != null){
                        for(int i = 0; i < args.length; i++, j++){
                            total[j] = args[i];
                        }
                    }

                    s.setThis(t);
                    try {
                        Object res = f.call(s, total);
                        return res;
                    }
                    finally {
                        s.clearThisNormal( null );
                    }
                }
            };
        }
    }
    /** @unexpose  */
    public static final JSFunction _functionBind = new Function_bind();

    /** Function.wrap: creates a wrapper function around this function
     *  See: http://www.prototypejs.org/api/function/wrap
     *  @param f  the function to wrap around this function
     */
    public static class Function_wrap extends JSFunctionCalls1 {
        public Object call( final Scope s , final Object f , final Object [] arguments){
            final JSFunction __method = (JSFunction)s.getThis();
            final JSFunction wrapper = (JSFunction)f;
            return new JSFunctionCalls0() {
                public Object call( Scope s , Object [] args ){
                    int totlen = 1;
                    if(args != null) totlen += args.length;
                    Object [] total = new Object [totlen];

                    Object tempthis = s.getThis();
                    s.setThis(__method);
                    try {
                        total[0] = _functionBind.call(s, tempthis, null);
                    }
                    finally {
                        s.clearThisNormal(null);
                    }

                    if(args != null){
                        for(int i = 0; i < args.length; i++){
                            total[i+1] = args[i];
                        }
                    }

                    //return new Integer(5);
                    return wrapper.call(s, total);
                }
            };

        }
    }
    /** @unexpose  */
    public static final JSFunction _functionWrap = new Function_wrap();

    public static class Class_create extends JSFunctionCalls0 {
        public Object call( Scope s, Object [] properties){
            int i = 0;
            Object parent = null;
            if(properties != null && properties.length > 0 && properties[0] instanceof JSFunction) {
                parent = properties[i++];
            }
            JSFunction klass = new JSFunctionCalls0 () {
                    public Object call( Scope s, Object [] extra){
                        JSObject t = (JSObject)s.getThis();
                        JSFunction init = (JSFunction)t.get("initialize");
                        init.call(s, extra);
                        return null;
                    }
                };
            klass.set("superclass", parent);
            // add subclasses attribute here

            if(parent != null) {
                JSFunction subclass = new JSFunctionCalls0 () {
                        public Object call( Scope s, Object [] extra){
                            return null;
                        }
                    };

                subclass.set("prototype", ((JSObject)parent).get("prototype"));
                klass.set("prototype", new JSObjectBase(subclass));
                // push subclass here
            }

            klass.set("addMethods", Prototype._classAddMethods);

            s.setThis(klass);
            if(properties != null)
                for(; i<properties.length; i++){
                    _classAddMethods.call(s, properties[i]);
                }
            s.clearThisNormal(true);

            JSObject prototype = (JSObject)klass.get("prototype");
            if(prototype.get("initialize") == null){
                prototype.set("initialize", new JSFunctionCalls0 () {
                        public Object call( Scope s, Object [] extra){
                            return null;
                        }
                    });
            }
            
            prototype.set("constructor", klass);

            return klass;

        }
    }

    /** @unexpose  */
    public static final JSFunction _classCreate = new Class_create();

    public static class Class_addMethods extends JSFunctionCalls1 {
        public Object call( Scope s, Object source, Object [] extra){
            JSObject t = (JSObject)s.getThis();
            Object anc = null;
            if(t.get("superclass") instanceof JSObject)
                anc = ((JSObject)t.get("superclass")).get("prototype");
            final Object ancestor = anc;
            JSObject jo = (JSObject)source;
            for(String property : jo.keySet()){
                Object value = jo.get(property);
                if (ancestor != null && value instanceof JSFunction){
                    JSFunction valf = (JSFunction)value;
                    Integer args = (Integer)valf.argumentNames().get("length");
                    if(args > 0 &&
                       ((JSFunction)value).argumentNames().getInt(0).equals("$super")) {
                        final String methodname = property;
                        final Object method = value;

                        value = new JSFunctionCalls0 () {
                                public Object call( Scope s, Object [] extra){
                                    JSFunction m = (JSFunction)((JSObject)ancestor).get(methodname);
                                    return m.call(s, extra);
                                }
                            };

                        s.setThis(value);
                        value = Prototype._functionWrap.call(s, method);

                        s.clearThisNormal(true);

                        ((JSObject)value).set("valueOf", new JSFunctionCalls0 () {
                                public Object call( Scope s, Object [] extra){
                                    return method;
                                }
                            });
                        ((JSObject)value).set("toString", new JSFunctionCalls0 () {
                                public Object call( Scope s, Object [] extra){
                                    return method.toString();
                                }
                            });
                    }
                }
                JSObject prototype = (JSObject)t.get("prototype");
                prototype.set(property, value);
            }
            return this;
        }
    }

    /** @unexpose  */
    public static final JSFunction _classAddMethods = new Class_addMethods();
    static {
	_classAddMethods.lock();
    }

    public static JSObjectBase newCopy(){
	JSObjectBase c = new JSObjectBase();
        
	JSObject m = new JSObjectBase();
        m.set("addMethods", _classAddMethods);
	
        c.set("Methods", m);
        c.set("create", _classCreate);

	return c;
    }

}

File: src/test/ed/js/prototype.js
// These are tests for the $A function, which I implemented but then decided
// wasn't very useful to us
/**
*    Copyright (C) 2008 10gen Inc.
*  
*    This program is free software: you can redistribute it and/or  modify
*    it under the terms of the GNU Affero General Public License, version 3,
*    as published by the Free Software Foundation.
*  
*    This program is distributed in the hope that it will be useful,
*    but WITHOUT ANY WARRANTY; without even the implied warranty of
*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*    GNU Affero General Public License for more details.
*  
*    You should have received a copy of the GNU Affero General Public License
*    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/*
a = $A([1, 2, 3]);
assert(a[0] == 1);
assert(a[1] == 2);
assert(a[2] == 3);

o = {0: 1, 4: 5, length: 5};
a = $A(o);

assert(a[0] == 1);
assert(a[4] == 5);
assert(a[2] == null);
*/

// Function.prototype.bind
f = function(a, b, c){
    return [this, a, b, c];
};

newf = f.bind(0, 1, 2);
g = newf(3);
assert(g[0] == 0);
assert(g[1] == 1);
assert(g[2] == 2);
assert(g[3] == 3);

// Make sure we accept empty arguments in either position
f2 = function(){
    return [this].concat(arguments);
};

newf2 = f2.bind(1);
g = newf2();
assert(g.length == 1);
assert(g[0] == 1);

g = newf2(2, 4);
assert(g.length == 3);
assert(g[0] == 1);
assert(g[1] == 2);
assert(g[2] == 4);


C = function(){
    this.x = 1;
};

C.prototype.getX = function(){
    return this.x;
};

C.prototype.getX = C.prototype.getX.wrap(
    function(proceed, factor){
        factor = (factor == null)? 1 : factor;
        return proceed()*factor;
    });

c1 = new C();
assert(c1.getX() == 1);
assert(c1.getX(5) == 5);
c1.x = 6;

assert(c1.getX() == 6);
assert(c1.getX(2) == 12);

a = function( a , b ){};
assert( a.argumentNames().length == 2 );
assert( a.argumentNames()[0] == "a" );
assert( a.argumentNames()[1] == "b" );

function b( a , b ){};
assert( b.argumentNames().length == 2 );
assert( b.argumentNames()[0] == "a" );
assert( b.argumentNames()[1] == "b" );


a = function(){};
a.prototype.m1 = function(i){
    return i * 2;
};

f = function(){};
f.superclass = a;

f.addMethods = Class.Methods.addMethods;

f.addMethods({
    m1: function($super, i){
        return $super(i+4);
    }
});

assert(f.m1(5) == 18);


// These unit tests are from Prototype -- check test/unit/base.html

var Animal = Class.create({
    initialize: function(name) {
        this.name = name;
    },
    name: "",
    eat: function() {
        return this.say("Yum!");
    },
    say: function(message) {
        return this.name + ": " + message;
    }
});

// subclass that augments a method
var Cat = Class.create(Animal, {
    eat: function($super, food) {
        if (food instanceof Mouse) return $super();
        else return this.say("Yuk! I only eat mice.");
    }
});

// empty subclass
var Mouse = Class.create(Animal, {});

assert(isFunction(Animal));
assert(Cat.superclass == Animal);
assert(Mouse.superclass == Animal);

var pet = new Animal("Nibbles");
assert(pet.name == "Nibbles");
assert(pet.say("Hi!") == "Nibbles: Hi!");
assert(Animal == pet.constructor);
assert(pet.superclass == null);

var Empty = Class.create();
assert('object', typeof new Empty);

var tom = new Cat("Tom");
assert(Cat == tom.constructor);
assert(Animal == tom.constructor.superclass);
assert("Tom" == tom.name);
assert("Tom: meow" == tom.say('meow'));
assert('Tom: Yuk! I only eat mice.' == tom.eat(new Animal));

assert('Tom: Yum!' == tom.eat(new Mouse));

// augment the constructor and test
var Dodo = Class.create(Animal, {
    initialize: function($super, name) {
        $super(name);
        this.extinct = true;
    },

    say: function($super, message) {
        return $super(message) + " honk honk";
    }
});

var gonzo = new Dodo('Gonzo');
assert('Gonzo' == gonzo.name);
assert(gonzo.extinct);
assert("Gonzo: hello honk honk" == gonzo.say("hello"));


var tom   = new Cat('Tom');
var jerry = new Mouse('Jerry');

Animal.addMethods({
    sleep: function() {
        return this.say('ZZZ');
    }
});

Mouse.addMethods({
    sleep: function($super) {
        return $super() + " ... no, can't sleep! Gotta steal cheese!";
    },
    escape: function(cat) {
        return this.say('(from a mousehole) Take that, ' + cat.name + '!');
    }
});

assertEqual = function(one, two, message){
    if (one != two) throw message;
}

assertUndefined = function(thing, message){
    if(! message) message = "not undefined";
    if(thing == null)
        ; // pass
    else
        throw message;
}

assertEqual('Tom: ZZZ', tom.sleep(), "added instance method not available to subclass");
assertEqual("Jerry: ZZZ ... no, can't sleep! Gotta steal cheese!", jerry.sleep());
assertEqual("Jerry: (from a mousehole) Take that, Tom!", jerry.escape(tom));
// insure that a method has not propagated *up* the prototype chain:
assertUndefined(tom.escape);
assertUndefined(new Animal().escape);

Animal.addMethods({
    sleep: function() {
        return this.say('zZzZ');
    }
});

assertEqual("Jerry: zZzZ ... no, can't sleep! Gotta steal cheese!", jerry.sleep());
